baybe/searchspace/discrete.py

```diff
@@ -255,16 +255,24 @@ def from_product(
         # Set defaults
         constraints = constraints or []
 
-        # Create a dataframe representing the experimental search space
-        exp_rep = parameter_cartesian_prod_to_df(parameters)
+        # Use Polars for Cartesian product of parameters and apply constraint
+        lazy_df = parameter_cartesian_prod_polars(parameters)
+        lazy_df = _apply_polars_constraint_filter(lazy_df, constraints)
+        pl_df = lazy_df.collect(streaming=True)
 
-        # Remove entries that violate parameter constraints
-        _apply_constraint_filter(exp_rep, constraints)
+        # Check if df is empty to prevent Polar's error
+        if pl_df.shape[0] == 0:
+            df = pd.DataFrame()
+        else:
+            df = pl_df.to_pandas()
+
+        # Remove entries that violate any remaining parameter constraints
+        _apply_pandas_constraint_filter(df, constraints)
```

Comment:
Now this applies the already used constraints **again**! This needs a cleverer logic

Reply:
yes, let me redesign it in th next PR where optionality is included because the lines here will be affected